[Screen Time Refactoring] Add API to show system Screen Time blocking view or blurred view instead

jesxilin opened the pull request:
#### a2d8d208f2cbb39feae772c55277edb794d8314d
<pre>
[Screen Time Refactoring] Add API to show system Screen Time blocking view or blurred view instead
<a href="https://bugs.webkit.org/show_bug.cgi?id=286845">https://bugs.webkit.org/show_bug.cgi?id=286845</a>
<a href="https://rdar.apple.com/140439133">rdar://140439133</a>

Reviewed by Aditya Keerthi.

Add API to give clients the choice to show WebKit&apos;s system Screen Time blocking view
or blur the web content.
- If true, WebKit will show the Screen Time blocking view.
- If false, WebKit shows a blurred view of the web content, which isn&apos;t interact-able
nor can any web content be actually visible.

Add three new API tests for this bit, as well as checking if the appropriate views are
added in the view hierarchy.

In WKWebView, WKSTVisualEffectView is a subclass NSVisualEffectView and we
override mouseDown among other events. This is so when the blurred view is up,
we do not allow navigation to links via clicking links
that are &quot;underneath&quot; the blurred view or other interactions such as clicking on
an input text field.
In other words, this is to make the web content with the blurred view
not interact-able.

* Source/WebKit/UIProcess/API/APIPageConfiguration.h:
(API::PageConfiguration::showsSystemScreenTimeBlockingView const):
(API::PageConfiguration::setShowsSystemScreenTimeBlockingView):
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKSTVisualEffectView mouseDown:]):
(-[WKSTVisualEffectView rightMouseDown:]):
(-[WKSTVisualEffectView mouseMoved:]):
(-[WKSTVisualEffectView mouseEntered:]):
(-[WKSTVisualEffectView mouseExited:]):
(-[WKWebView _installScreenTimeWebpageController]):
(-[WKWebView observeValueForKeyPath:ofObject:change:context:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
(-[WKWebViewConfiguration _showsSystemScreenTimeBlockingView]):
(-[WKWebViewConfiguration _setShowsSystemScreenTimeBlockingView:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfigurationPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
(blurredViewIsPresent):
(systemScreenTimeBlockingViewIsPresent):
(testShowsSystemScreenTimeBlockingView):
(TEST(ScreenTime, ShowSystemScreenTimeBlockingTrue)):
(TEST(ScreenTime, ShowSystemScreenTimeBlockingFalse)):
(TEST(ScreenTime, ShowSystemScreenTimeBlockingFalseAndRemoved)):

Canonical link: <a href="https://commits.webkit.org/289956@main">https://commits.webkit.org/289956@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/183ba7ee79aa95cd0fec84e343ecfad4d15994b0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/88487 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/8006 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/42927 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/93445 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/39241 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/90538 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/8393 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/16190 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/68249 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/25967 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/91489 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/6426 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/80037 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/48616 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/6198 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/34453 "Found 3 new test failures: fonts/monospace.html fonts/sans-serif.html fonts/serif.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/38349 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/76571 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/35336 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/95287 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/15662 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/11478 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/77112 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/15918 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/75893 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/76375 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/20767 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/19184 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/8701 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13841 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/15678 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/20986 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/15419 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/18868 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/17201 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -234,10 +234,41 @@
+- (void)rightMouseDown:(NSEvent *)event;
+- (void)mouseMoved:(NSEvent *)event;
+- (void)mouseEntered:(NSEvent *)event;
+- (void)mouseExited:(NSEvent *)event;
```
No need to declare these methods.

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -234,10 +234,41 @@
+@implementation WKSTVisualEffectView
+- (void)mouseDown:(NSEvent *)event
+{
+    return;
```
`return` is unnecessary.

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -403,14 +437,38 @@ - (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(N
+        if (oldIsBlockedByScreenTime != _isBlockedByScreenTime) {
+            if (!_screenTimeBlurredSnapshot && [_configuration _shouldSuppressScreenTimeBlockingView]) {
+#if PLATFORM(MAC)
+                _screenTimeBlurredSnapshot = [[WKSTVisualEffectView alloc] init];
```
This is a memory leak (need `adoptNS`).

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -403,14 +437,38 @@ - (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(N
+                [_screenTimeBlurredSnapshot setBlendingMode:NSVisualEffectBlendingModeWithinWindow];
+#else
+                UIBlurEffect *blurEffect = [UIBlurEffect effectWithStyle:UIBlurEffectStyleSystemUltraThinMaterial];
+                    _screenTimeBlurredSnapshot = [[UIVisualEffectView alloc] initWithEffect:blurEffect];
```
Leak, and indentation.

pxlcoder reviewed (commented):
> appropiate

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -403,14 +437,38 @@ - (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(N
     if (context == &screenTimeWebpageControllerBlockedKVOContext) {
         BOOL urlWasBlocked = dynamic_objc_cast<NSNumber>(change[NSKeyValueChangeOldKey]).boolValue;
         BOOL urlIsBlocked = dynamic_objc_cast<NSNumber>(change[NSKeyValueChangeNewKey]).boolValue;
+        BOOL oldIsBlockedByScreenTime = _isBlockedByScreenTime;
```
`wasBlockedByScreenTime`?

jesxilin replied:
Idk but I thought it was sounded a bit weird without a subject like `urlWasBlocked` so I opted for "old" instead. But I will change it!

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -403,14 +437,38 @@ - (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(N
+                [_screenTimeBlurredSnapshot setMaterial:NSVisualEffectMaterialUnderWindowBackground];
+                [_screenTimeBlurredSnapshot setBlendingMode:NSVisualEffectBlendingModeWithinWindow];
+#else
+                UIBlurEffect *blurEffect = [UIBlurEffect effectWithStyle:UIBlurEffectStyleSystemUltraThinMaterial];
```
```suggestion
                RetainPtr blurEffect = [UIBlurEffect effectWithStyle:UIBlurEffectStyleSystemUltraThinMaterial];
```

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -403,14 +437,38 @@ - (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(N
+                    [[_screenTimeBlurredSnapshot leadingAnchor] constraintEqualToAnchor:self.leadingAnchor],
+                    [[_screenTimeBlurredSnapshot topAnchor] constraintEqualToAnchor:self.topAnchor]
+                ]];
+
```
```suggestion
```

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -234,10 +234,31 @@
-static void *screenTimeWebpageControllerBlockedKVOContext = &screenTimeWebpageControllerBlockedKVOContext;
+#if PLATFORM(MAC)
+@interface WKSTVisualEffectView : NSVisualEffectView
+@end
```
```suggestion
@end

```

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -234,10 +234,31 @@
+#if PLATFORM(MAC)
+@interface WKSTVisualEffectView : NSVisualEffectView
+@end
+@implementation WKSTVisualEffectView
```
```suggestion
@implementation WKSTVisualEffectView

```

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -234,10 +234,31 @@
+@implementation WKSTVisualEffectView
+- (void)mouseDown:(NSEvent *)event
+{
+}
```
Nit: Add a newline after each method.
```suggestion
}

```

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -256,4 +299,18 @@ - (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(N
+TEST(ScreenTime, ShouldSuppressBlockingViewTrueAndRemoved)
+{
+    testSuppressScreenTimeBlocking(true, true);
+}
```
```suggestion
}

```

aprotyas commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -121,6 +121,49 @@ - (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(N
+    return false;
+}
+
+static void testSuppressScreenTimeBlocking(bool suppressScreenTimeBlocking, bool isRemoveTest = false)
```
Optional: Boolean trap. Makes the call sites hard to follow. https://webkit.org/code-style-guidelines/#names-enum-to-bool

```suggestion
enum class SuppressScreenTimeBlocking : bool { No, Yes };
enum class TestBlockingViewRemoval : bool { No, Yes };

static void testSuppressScreenTimeBlocking(SuppressScreenTimeBlocking suppressScreenTimeBlocking, TestBlockingViewRemoval testBlockingViewRemoval = TestBlockingViewRemoval::No)
```

pxlcoder commented on Source/WebKit/UIProcess/API/APIPageConfiguration.h:
```diff
@@ -387,6 +387,9 @@ class PageConfiguration : public ObjectImpl<Object::Type::PageConfiguration> {
     void setIncompleteImageBorderEnabled(bool enabled) { m_data.incompleteImageBorderEnabled = enabled; }
 
+    bool showsSystemScreenTimeBlockingView() const { return m_data.showsSystemScreenTimeBlockingView; }
+    void setshowsSystemScreenTimeBlockingView(bool shows) { m_data.showsSystemScreenTimeBlockingView = shows; }
```
```suggestion
    void setShowsSystemScreenTimeBlockingView(bool shows) { m_data.showsSystemScreenTimeBlockingView = shows; }
```

jesxilin replied:
Woah.

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -372,14 +400,17 @@ - (void)_installScreenTimeWebpageController
         [_screenTimeWebpageController setSuppressUsageRecording:![_configuration websiteDataStore].isPersistent];
 
         RetainPtr screenTimeView = [_screenTimeWebpageController view];
```
I would leave a comment about why this is created, regardless of `[_configuration _showsSystemScreenTimeBlockingView]`.

jesxilin replied:
Got it!

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfigurationPrivate.h:
```diff
@@ -95,6 +95,7 @@ typedef NS_ENUM(NSUInteger, _WKContentSecurityPolicyModeForExtension) {
 @property (nonatomic, setter=_setIncompleteImageBorderEnabled:) BOOL _incompleteImageBorderEnabled WK_API_AVAILABLE(macos(10.14), ios(12.0));
 @property (nonatomic, setter=_setDrawsBackground:) BOOL _drawsBackground WK_API_AVAILABLE(macos(10.14), ios(12.0));
 @property (nonatomic, setter=_setShouldDeferAsynchronousScriptsUntilAfterDocumentLoad:) BOOL _shouldDeferAsynchronousScriptsUntilAfterDocumentLoad WK_API_AVAILABLE(macos(10.14), ios(12.0));
+@property (nonatomic, setter=_setshowsSystemScreenTimeBlockingView:) BOOL _showsSystemScreenTimeBlockingView WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA));
```
```suggestion
@property (nonatomic, setter=_setShowsSystemScreenTimeBlockingView:) BOOL _showsSystemScreenTimeBlockingView WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA));
```

jesxilin replied:
Yep!

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
```diff
@@ -730,6 +730,16 @@ - (void)_setAllowUniversalAccessFromFileURLs:(BOOL)allowUniversalAccessFromFileU
+    return _pageConfiguration->showsSystemScreenTimeBlockingView();
+}
+
+- (void)_setshowsSystemScreenTimeBlockingView:(BOOL)shows
```
```suggestion
- (void)_setShowsSystemScreenTimeBlockingView:(BOOL)shows
```

jesxilin replied:
Changed!

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -256,4 +310,25 @@ - (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(N
+    EXPECT_FALSE([[webView configuration] _showsSystemScreenTimeBlockingView]);
+    // Check if blurred blocking view is removed when URLIsBlocked is false.
+    EXPECT_FALSE(viewIsPresent(webView.get()));
+    EXPECT_FALSE(viewIsPresent(webView.get()));
```
Should the second check be for `systemScreenTimeBlockingViewIsPresent`? `viewIsPresent` is duplicated.

jesxilin replied:
No, it's supposed to be one check... I just changed it. Will push!

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -121,6 +121,60 @@ - (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(N
+    return false;
+}
+
+static BOOL systemScreenTimeBlockingViewIsPresent(TestWKWebView *webView, STWebpageController *controller)
```
You don't need to pass in the `STWebpageController`, since it's accessible from `webView`. But not a big deal either.

jesxilin replied:
I'll access it from webView!

jesxilin committed:
[Screen Time Refactoring] Add API to show system Screen Time blocking view or blurred view instead
https://bugs.webkit.org/show_bug.cgi?id=286845
rdar://140439133

Reviewed by Aditya Keerthi.

Add API to give clients the choice to show WebKit's system Screen Time blocking view
or blur the web content.
- If true, WebKit will show the Screen Time blocking view.
- If false, WebKit shows a blurred view of the web content, which isn't interact-able
nor can any web content be actually visible.

Add three new API tests for this bit, as well as checking if the appropriate views are
added in the view hierarchy.

In WKWebView, WKSTVisualEffectView is a subclass NSVisualEffectView and we
override mouseDown among other events. This is so when the blurred view is up,
we do not allow navigation to links via clicking links
that are "underneath" the blurred view or other interactions such as clicking on
an input text field.
In other words, this is to make the web content with the blurred view
not interact-able.

* Source/WebKit/UIProcess/API/APIPageConfiguration.h:
(API::PageConfiguration::showsSystemScreenTimeBlockingView const):
(API::PageConfiguration::setShowsSystemScreenTimeBlockingView):
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKSTVisualEffectView mouseDown:]):
(-[WKSTVisualEffectView rightMouseDown:]):
(-[WKSTVisualEffectView mouseMoved:]):
(-[WKSTVisualEffectView mouseEntered:]):
(-[WKSTVisualEffectView mouseExited:]):
(-[WKWebView _installScreenTimeWebpageController]):
(-[WKWebView observeValueForKeyPath:ofObject:change:context:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
(-[WKWebViewConfiguration _showsSystemScreenTimeBlockingView]):
(-[WKWebViewConfiguration _setShowsSystemScreenTimeBlockingView:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfigurationPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
(blurredViewIsPresent):
(systemScreenTimeBlockingViewIsPresent):
(testShowsSystemScreenTimeBlockingView):
(TEST(ScreenTime, ShowSystemScreenTimeBlockingTrue)):
(TEST(ScreenTime, ShowSystemScreenTimeBlockingFalse)):
(TEST(ScreenTime, ShowSystemScreenTimeBlockingFalseAndRemoved)):

Canonical link: https://commits.webkit.org/289956@main
